### PR TITLE
Automatically generate solution filters

### PR DESCRIPTION
**Pull request by jamescrosswell**
#skip-changelog

# Overview

First step towards addressing https://github.com/getsentry/sentry-dotnet/issues/2560

This PR has two goals:
1. Remove multiple `*.sln` files. The idea is that we would only maintain `Sentry.sln` moving forward and this would contain all of the packages/samples/tests required to build sentry-dotnet on any machine
2. Automate the generation and maintenance of the various `*.slnf` solution filter files. Instead of manually adding projects to each of these, every time we create new packages, samples etc., we can configure which projects to include in filters using glob includes/excludes in [scripts/generate-solution-filters-config.yaml](https://github.com/getsentry/sentry-dotnet/pull/2576/files#diff-80a5e89327c637e5542ddb0f4ffd1d62cfb71524258e5c1ec9c3f00ca781d81f)

## Note on build checks

There was some [discussion about whether the solution filters should be committed to source control](https://github.com/getsentry/sentry-dotnet/pull/2576#issuecomment-1699882565) or not. Current behavriour is:
- Solution Filters should be committed to GIT
- When running builds on CI we
  - Run the script to regenerate solution filters
  - Perform a dirty check to ensure none of the filters has changed

This guarantees that the solution filters committed to source control are consistent with the solution filter generation script... you can't change one without changing the other. 

# TODO in follow up pull requests

This PR minimizes changes to the build process and only aims to alter (very slightly) solution filters where it seems clear these include/exclude files that they shouldn't do.

Once we're happy with this PR, we probably want to follow up with a couple more to rationalise `generate-solution-filters-config.yaml` file and remove unecessary solution filters.

### Tidying generate-solution-filters-config.yaml

The includes and excludes configured for solution filters in `generate-solution-filters-config.yaml` currently are all pretty bespoke... There's no obvious logic to what gets included in each filter and practically no comments explaining why some things are included and others excluded. 

Ideally, we invest a little time describing (in comments in the config) what the purpose of each solution filter is for and therefore why some files are deliberately included or excluded... and hopefully, from that, we could start to tease out logical groups that could be defined in the config, that would be reused in the includes and excludes across solution filters. Currently the config is making fairly minimal use of groups (which is a pity, since these could help a lot with readibility).

### Removing unecessary solution filters

1. Get rid of `Sentry-CI-Pack.slnf`... there's no real need for this as all of our test and sample projects can be excluded from packing via the use of `<IsPackable>false</IsPackable>` in `Directory.Build.props` files. If we're building and testing everything for release using `Sentry-CI-Build-macOS.slnf`, we should be able to use that solution filter for packing as well.
2. Potentially we could get rid of `Sentry-CI-Test.slnf` and `Sentry-CI-CodeQL.slnf` by the same logic

### CODE REVIEW

**Review by bitsandfoxes — APPROVED** [review]
I really like where this is going! One solution to rule them all!
So by the end of this restructuring, there will be no checked in `.slnf`?

File: scripts/generate-solution-filters.ps1 [hunk 1]
```diff
@@ -0,0 +1,102 @@
+param(
+    [string]$ConfigFile = "generate-solution-filters-config.yaml"
+)
+
```
Comment (vaind) [inline scripts/generate-solution-filters.ps1:4]:
```suggestion

Set-StrictMode -Version Latest
$ErrorActionPreference = "Stop"

```

File: src/Sentry.Profiling/Sentry.Profiling.csproj [hunk 2]
```diff
@@ -4,6 +4,8 @@
     <!-- TODO check and update the list of supported frameworks. -->
     <TargetFrameworks>net6.0</TargetFrameworks>
     <PackageTags>$(PackageTags);Profiling;Diagnostic</PackageTags>
+    <!-- TODO: re-enable packing once we've resolved all dependencies for Sentry.Profiling -->
+    <IsPackable>false</IsPackable>
```
[the remark this answers is not in the document]
Comment (jamescrosswell) [reply src/Sentry.Profiling/Sentry.Profiling.csproj:8]: That's a question for @vaind I think...

[the remark this answers is not in the document]
Comment (vaind) [reply src/Sentry.Profiling/Sentry.Profiling.csproj:8]: This is about the perfview dependency and more importantly, it's transitive dependencies on some native precompiled stuff that's not available on nuget but needs a separate package source definition.

**Pull request merged**